InnerEye/ML/Histopathology/datamodules/base_module.py

```diff
@@ -87,17 +100,20 @@ def prepare_data(self) -> None:
             self._load_dataset(self.val_dataset, stage='val', shuffle=True)
             self._load_dataset(self.test_dataset, stage='test', shuffle=True)
 
-    def _dataset_pickle_path(self, stage: str) -> Optional[Path]:
+    def _dataset_pt_path(self, stage: str) -> Optional[Path]:
         if self.cache_dir is None:
             return None
-        return self.cache_dir / f"{stage}_dataset.pkl"
+        return self.cache_dir / f"{stage}_dataset.pt"
 
     def _load_dataset(self, tiles_dataset: TilesDataset, stage: str, shuffle: bool) -> Dataset:
-        dataset_pickle_path = self._dataset_pickle_path(stage)
+        dataset_pt_path = self._dataset_pt_path(stage)
 
-        if dataset_pickle_path and dataset_pickle_path.exists():
-            with dataset_pickle_path.open('rb') as f:
-                return pickle.load(f)
+        if dataset_pt_path and dataset_pt_path.exists():
+            # torch.load will reload on GPU by default, same device it was saved from
+            memory_location = torch.device('cpu') if self.save_precache == CacheLocation.CPU else None
```

Comment:
two reasons:
1 - no need to, torch.load will reload on GPU by default (see comment above)
2 - we would need to specify the device, I understand the correct syntax is `cuda:device`, that means I would need to propagate the device info here

Reply:
1. It will reload on whatever original device. Note that this class is generic w.r.t. the `transform`, meaning the data to cache might live in CPU. In this case, the user who specified `CacheLocation.GPU` would be surprised that their data is not loaded on GPU.
2. Just `'cuda'` without an index maps to the default GPU: https://pytorch.org/docs/stable/notes/cuda.html

Reply:
1. I hear you, I haven't planned for the option where the result of the transformation is not already on gpu. Let me capture that case too.
2. I know we can use `cuda` for a generic device but here the saving will happen on one single device and when we reload we want each process to reload in a different device, if we just specify `cuda` I understand it will instead reload always to the same, whatever device was used before saving. [https://pytorch.org/docs/stable/generated/torch.load.html](https://pytorch.org/docs/stable/generated/torch.load.html).  Am I wrong?

Reply:
1. Sounds good.
2. You're right. Maybe a more flexible and robust option is to let the user directly specify the device (`'cpu'`, `'cuda:xxx'`, or `None`)?

Reply:
Discussed at voice. In the end I chose not to support the option save in cpu - load to gpu. The new variable name should convey clearly to the user what are the possibilities.